File: branches/with-libpq-backend/src/dialogs/DatasetEditorDialog.cpp
/* 
 * File:   DatasetEditorDialog.cpp
 * Author: jolo
 * 
 * Created on 10. Dezember 2009, 21:22
 */

#include "DatasetEditorDialog.h"

#include <QDialog>

#include <QLayout>
#include <QDialogButtonBox>
#include <QGroupBox>
#include <QTabWidget>
#include <QApplication>

#include "IdLabel.h"
#include "NameEdit.h"
#include "DescriptionEdit.h"
#include "Dataset.h"
#include "CustomSymbolInBedItemModel.h"
#include "ProfileLogger.h"
#include "ProfileLoggerDatabase.h"
#include "Postgres.h"
#include "DataManager.h"

DatasetEditorDialog::DatasetEditorDialog(QWidget* p, Dataset* d)
  : QDialog(p),
    r(0),
    lC(0),
    wC(1),
    _data(d),
    _dataManager(0),
    _postgres(0),
    _database(0),
    _tabW(0),
    _idW(0),
    _mainPageW(0),
    _nameW(0),
    _descriptionW(0),
    _bbW(0) {
  _postgres = (static_cast<ProfileLogger*>(QApplication::instance()))->getPostgres();
  _database = (static_cast<ProfileLogger*>(QApplication::instance()))->getProfileLoggerDatabase();
  setLayout(new QVBoxLayout());
  _tabW = new QTabWidget(this);
  layout()->addWidget(_tabW);
  connect(this, SIGNAL(showDatasetRequest(Dataset*)), this, SLOT(slotShowDataset(Dataset*)));
}

DatasetEditorDialog::~DatasetEditorDialog() {
}

QWidget* DatasetEditorDialog::addMainPage(const QString& title) {
  _mainPageW = new QGroupBox(_tabW);
  _mainPageW->setLayout(new QGridLayout());
  _tabW->addTab(_mainPageW, title);
  return _mainPageW;
}

IdLabel* DatasetEditorDialog::addIdLabel() {
  _idW = new IdLabel(_mainPageW);
  ((QGridLayout*) (_mainPageW->layout()))->addWidget(_idW->getLabel(), r, lC);
  ((QGridLayout*) (_mainPageW->layout()))->addWidget(_idW, r, wC);
  r++;
  return _idW;
}

NameEdit* DatasetEditorDialog::addNameEdit() {
  _nameW = new NameEdit(_mainPageW);
  connect(_nameW, SIGNAL(nameChanged(const QString&)), this, SLOT(slotNameChanged(const QString&)));
  ((QGridLayout*) (_mainPageW->layout()))->addWidget(_nameW->getLabel(), r, lC);
  ((QGridLayout*) (_mainPageW->layout()))->addWidget(_nameW, r, wC);
  r++;
  return _nameW;
}

DescriptionEdit* DatasetEditorDialog::addDescriptionEdit() {
  _descriptionW = new DescriptionEdit(_mainPageW);
  connect(_descriptionW, SIGNAL(descriptionChanged(const QString&)), this, SLOT(slotDescriptionChanged(const QString&)));
  ((QGridLayout*) (_mainPageW->layout()))->addWidget(_descriptionW->getLabel(), r, lC);
  ((QGridLayout*) (_mainPageW->layout()))->addWidget(_descriptionW, r, wC);
  r++;
  return _descriptionW;
}

QDialogButtonBox* DatasetEditorDialog::addButtons() {
  _bbW = new QDialogButtonBox(QDialogButtonBox::Ok | QDialogButtonBox::Cancel,
			      Qt::Horizontal,
			      this);
  connect(_bbW, SIGNAL(rejected()), this, SLOT(reject()));
  connect(_bbW, SIGNAL(accepted()), this, SLOT(accept()));
  layout()->addWidget(_bbW);
  return _bbW;
}

void DatasetEditorDialog::slotNameChanged(const QString& s) {
  if (_data) {
    _data->setName(s);
  }
}

void DatasetEditorDialog::slotDescriptionChanged(const QString& s) {
  if (_data) {
    _data->setDescription(s);
  }
}

void DatasetEditorDialog::slotShowDataset(Dataset* d) {
  _data = d;

  if (!_data) {
    clear();
    return;
  }

  if (_idW) {
    _idW->setId(_data->getId());
  }

  if (_nameW) {
    _nameW->setText(_data->getName());
  }

  if (_descriptionW) {
    _descriptionW->setText(_data->getDescription());
  }
}

void DatasetEditorDialog::clear() {
  if (_idW) {
    _idW->setId(0);
  }

  if (_nameW) {
    _nameW->clear();
  }

  if (_descriptionW) {
    _descriptionW->clear();
  }
}

void DatasetEditorDialog::begin() {
  getPostgres()->begin();
}

void DatasetEditorDialog::commit() {
  getPostgres()->commit();
}

void DatasetEditorDialog::rollback() {
  getPostgres()->rollback();
}

void DatasetEditorDialog::reject() {
  done(QDialog::Rejected);
}

void DatasetEditorDialog::accept() {
  done(QDialog::Accepted);
}

File: branches/experimental/py/Gui/Dialogs/DatasetEditorDialog.py
from Dialog import Dialog

from PyQt4.QtGui import *
from PyQt4.QtCore import *

from Gui.Widgets.IdLabel import IdLabel
from Gui.Widgets.NameEdit import NameEdit
from Gui.Widgets.DescriptionEdit import DescriptionEdit
from Gui.Widgets.IntLineEdit import IntLineEdit
from Gui.Widgets.SVGLoaderWidget import SVGLoaderWidget

from Gui.Dialogs.DatabaseExceptionDialog import DatabaseExceptionDialog

from sqlalchemy.exc import *
from sqlalchemy.orm.exc import *

class DatasetEditorDialog(Dialog):
    def __init__(self, parent, data):
        Dialog.__init__(self, parent)
        self.data = data
        self.setLayout(QVBoxLayout(self))
    def addContentPanel(self, title):
        self.contentW = QGroupBox(title, self)
        self.contentW.setLayout(QGridLayout(self.contentW))
        self.layout().addWidget(self.contentW)
        self.currentContentRow = 0
        self.labelCol = 0
        self.widgetCol = 1
    def addButtons(self):
        self.bbW = QDialogButtonBox(QDialogButtonBox.Save | QDialogButtonBox.Cancel,
                                    Qt.Horizontal,
                                    self)
        self.bbW.accepted.connect(self.accept)
        self.bbW.rejected.connect(self.reject)
        self.layout().addWidget(self.bbW)
    def addLabelWidgetPair(self, labelW, widget):
        self.contentW.layout().addWidget(labelW, self.currentContentRow, self.labelCol)
        self.contentW.layout().addWidget(widget, self.currentContentRow, self.widgetCol)
        self.currentContentRow += 1
    def addIdDisplay(self):
        self.idL = self.createOneLineLabel(self.tr("ID"))
        self.idW = IdLabel(self.contentW)
        self.addLabelWidgetPair(self.idL, self.idW)
    def addIntEdit(self, label, min=1, max=999999):
        self.intL = self.createOneLineLabel(label)
        self.intW = IntLineEdit(self)
        self.intL.setBuddy(self.intW)
        self.intW.validator().setRange(min, max)
        self.addLabelWidgetPair(self.intL, self.intW)
    def addNameEdit(self, label=None):
        lbl = label
        if lbl is None:
            lbl = self.tr("&Name")
        self.nameL = self.createOneLineLabel(lbl)
        self.nameW = NameEdit(self.contentW)
        self.nameL.setBuddy(self.nameW)
        self.addLabelWidgetPair(self.nameL, self.nameW)
    def addDescriptionEdit(self, label=None):
        lbl = label
        if lbl is None:
            lbl = self.tr("&Description")
        self.descriptionL = self.createMultiLineLabel(lbl)
        self.descriptionW = DescriptionEdit(self.contentW)
        self.descriptionL.setBuddy(self.descriptionW)
        self.addLabelWidgetPair(self.descriptionL, self.descriptionW)
    def addSVGLoader(self, label=None):
        lbl = label
        if lbl is None:
            lbl = self.tr("&SVG File")
        self.svgLoaderL = self.createOneLineLabel(lbl)
        self.svgLoaderW = SVGLoaderWidget(self.contentW)
        self.svgLoaderL.setBuddy(self.svgLoaderW)
        self.addLabelWidgetPair(self.svgLoaderL, self.svgLoaderW)
    def createOneLineLabel(self, lbl):
        lbl = QLabel(lbl, self.contentW)
        lbl.setAlignment(Qt.AlignVCenter | Qt.AlignRight)
        return lbl
    def createMultiLineLabel(self, lbl):
        lbl = QLabel(lbl, self.contentW)
        lbl.setAlignment(Qt.AlignTop | Qt.AlignRight)
        return lbl
    def validate(self):
        return True
    def accept(self):
        if not self.validate():
            return
        try:
            if not self.data.hasId():
                QApplication.instance().db.session.add(self.data)
            QApplication.instance().db.session.commit()
            self.done(QDialog.Accepted)
        except IntegrityError, e:
            dlg = DatabaseExceptionDialog(QApplication.activeWindow(), e)
            dlg.exec_()
            QApplication.instance().db.session.rollback()
        except ConcurrentModificationError, e:
            dlg = DatabaseExceptionDialog(QApplication.activeWindow(), e)
            dlg.exec_()
            QApplication.instance().db.session.rollback()
        except ProgrammingError, e:
            dlg = DatabaseExceptionDialog(QApplication.activeWindow(), e)
            dlg.exec_()
            QApplication.instance().db.session.rollback()

    def reject(self):
        if self.data.hasId():
            QApplication.instance().db.session.refresh(self.data)
        self.done(QDialog.Rejected)
